src/DockWidgetBase.cpp

```diff
@@ -543,7 +543,7 @@ void DockWidgetBase::onClosed(QCloseEvent *e)
         d->close();
 }
 
-DockWidgetBase *DockWidgetBase::deserialize(LayoutSaver::DockWidget::Ptr saved)
+DockWidgetBase *DockWidgetBase::deserialize(const LayoutSaver::DockWidget::Ptr &saved)
```

Comment:
Passing shared pointers by const-ref is bad, as it inhibits the ref-count:

Evil example #1:
`

void MyClass::foo(const Ptr &ptr)
{
    if (!ptr) return; // it has ref-count=1, held somewhere else
    doComplexThings(); // which might lead to make the pointer decrease ref-count (if it's a member of a deleted class)
    ptr->bar(); // oops, dereferencing a null pointer
}
`

Which defeats the purpose of shared pointers, which is to keep tabs on resources

Reply:
Ofc, the example is contrived, but I've seen it happen on decent code, and wouldn't feel comfortable changing a huge code base

Reply:
This same reasoning is applicable to COW containers like Qt's - I've seen the same happening there. So either you also never pass them via `const&` or you do but then also pass ptrs by `const&` too and fix the issues you find via ASAN/UBSAN. Such shadowing issues are rare thankfully.

Reply:
Yeah, ok, I'll buy that :)